core/src/main/java/org/elasticsearch/repositories/blobstore/ChecksumBlobStoreFormat.java

```
@@ -0,0 +1,217 @@
+            byte[] bytes = ByteStreams.toByteArray(inputStream);
+            final String resourceDesc = "ChecksumBlobStoreFormat.readBlob(blob=\"" + blobName + "\")";
+            try (ByteArrayIndexInput indexInput = new ByteArrayIndexInput(resourceDesc, bytes)) {
+                CodecUtil.checksumEntireFile(indexInput);
+                CodecUtil.checkHeader(indexInput, codec, VERSION, VERSION);
+                long filePointer = indexInput.getFilePointer();
+                long contentSize = indexInput.length() - CodecUtil.footerLength() - filePointer;
+                BytesReference bytesReference = new BytesArray(bytes, (int) filePointer, (int) contentSize);
+                return read(bytesReference);
+            } catch (CorruptIndexException | IndexFormatTooOldException | IndexFormatTooNewException ex) {
+                // we trick this into a dedicated exception with the original stacktrace
+                throw new CorruptStateException(ex);
+            }
+        }
+    }
+
+    /**
+     * Writes blob in atomic manner with resolving the blob name using {@link #blobName} and {@link #tempBlobName} methods.
+     * <p/>
+     * The blob will be compressed and checksum will be written if required.
+     *
+     * Atomic move might be very inefficient on some repositories. It also cannot override existing files.
+     *
+     * @param obj           object to be serialized
+     * @param blobContainer blob container
+     * @param name          blob name
+     * @throws IOException
+     */
+    public void writeAtomic(T obj, BlobContainer blobContainer, String name) throws IOException {
+        String blobName = blobName(name);
+        String tempBlobName = tempBlobName(name);
+        writeBlob(obj, blobContainer, tempBlobName);
+        try {
+            blobContainer.move(tempBlobName, blobName);
+        } catch (IOException ex) {
+            // Move failed - try cleaning up
+            blobContainer.deleteBlob(tempBlobName);
```

Maybe we should inject a logger so we can log something here? Otherwise we silently fail the writeAtomic?